src/softerllama/modeling_softerllama.py

```diff
@@ -224,12 +194,9 @@ def forward(
 
         # upcast attention to fp32
         # softermax overriding original attention softmax
-        attn_weights = softermax(
-            attn_weights, self.n_bias, dim=-1, dtype=torch.float32
-        ).to(query_states.dtype)
-        attn_weights = nn.functional.dropout(
-            attn_weights, p=self.attention_dropout, training=self.training
-        )
+        attn_weights = softermax(attn_weights, self.n_bias, dim=-1).type(dtype=torch.float32).to(query_states.dtype)
+        attn_weights = clip_logits(attn_weights, self.n_clip, dim=-1)
```

Comment:
`clip` stacks on `softermax` -- to turn off clip we just set `n_clip=0`, to turn off softermax we just set `n_bias=0`.

@Clyde013 Is this inefficient? Or do CUDA optimisations / fused kernels negate that at the limit

Reply:
Realistically the inefficiencies are going to be incredibly minor overhead, and if it does turn out to be a big deal we can wrap the whole thing inside an if-statement. Not a big enough deal for us to care too much about it.